File: tarea2/pizza2.cpp
#include <bits/stdc++.h>
#include <math.h>
#include <algorithm>
#include <iostream>
#include <vector>
using ll = long long;
using namespace std;

int N, S;
ll ts = 0;
vector<vector<ll>> vals;

ll h(ll n1, ll tot, ll s) {
  ll cnt1 = n1;
  ll cnt2 = tot - n1;
  ll res = 0;
  ll i = 0;
  while (cnt1 > 0 && i < s) {
    if (cnt1 - vals[i][1] >= 0) {
      res += vals[i][1] * vals[i][2];
      cnt1 -= vals[i][1];
    } else {
      res += cnt1 * vals[i][2];
      res += (vals[i][1] - cnt1) * vals[i][3];
      cnt2 -= vals[i][1] - cnt1;
      cnt1 = 0;
    }
    i += 1;
  }
  while (cnt2 > 0 && i < s) {
    res += vals[i][1] * vals[i][3];
    cnt2 -= vals[i][1];
    i += 1;
  }
  return res;
}

int main() {
  cin >> N;
  cin >> S;
  for (int i = 0; i < N; i++) {
    int s, a, b;
    cin >> s;
    cin >> a;
    cin >> b;
    ts += s;
    vector<ll> tmp;
    tmp.push_back(a - b);
    tmp.push_back(s);
    tmp.push_back(a);
    tmp.push_back(b);
    vals.push_back(tmp);
  }
  ll p = ceil((float)ts / S);
  // cout << slices << endl;
  ll dummy = S * p - ts;
  if (ts > S) {
    vector<ll> dv;
    dv.push_back(0);
    dv.push_back(dummy);
    dv.push_back(0);
    dv.push_back(0);
    vals.push_back(dv);
  }
  sort(vals.begin(), vals.end(),
       [](const vector<ll>& a, const vector<ll>& b) { return a[0] > b[0]; });

  // filtrar por numero de tortas
  //vector<ll> posible;
  //for (ll i = 0; i < S * p + 1; i++) {
  //  if (ceil((float)i / S) + ceil((float)(S * p - i) / S) <= p) {
  //    posible.push_back(i);
  //  }
 // }

  ll res = 0;
  ll l = 0;
  ll r = p;
  ll sz = vals.size();
  while (l < r) {
    ll mid = (l + r) / 2;
    ll ch = h(mid*S, S * p, sz);
    ll nh = h((mid + 1)*S, S * p, sz);
    if (nh - ch > 0) {
      l = mid + 1;
    } else {
     r = mid;
    }
  }
  cout << h(l*S, S*p, sz) << endl;
}

File: tarea2/pizza2.py
import sys
from math import ceil
def get_ints(): return map(int, sys.stdin.readline().strip().split())


N, S = get_ints()
vals = []
ts = 0
for i in range(N):
    s, a, b = get_ints()
    ts += s
    vals.append([a-b, s, a, b])
p = ceil(ts/S)
dummy = S*p-ts
# trozos sobrantes - mas de 1 pizza
if ts > S:
    vals.append([0, dummy, 0, 0])
# ordenar de a-b mayor a menor
vals.sort(reverse=True)


def h(n1):
    cnt1 = n1
    cnt2 = p*S-n1
    res = 0
    # comer n1 hasta que se acabe
    i = 0
    while(cnt1 > 0):
       # print(i)
       # print(cnt1)
        if cnt1 - vals[i][1] >= 0:
            res += vals[i][1]*vals[i][2]
           # print(res)
            cnt1 -= vals[i][1]
        else:
            # termino de comer lo que me queda de n1
            res += cnt1*vals[i][2]
           # print("termino de comer n1")
           # print(res)
            # el resto vals[i][1] como de n2
           # print("el resto como n2")
            res += (vals[i][1]-cnt1)*vals[i][3]
           # print(res)
            cnt2 -= (vals[i][1]-cnt1)
            cnt1 = 0
        i += 1
    while(cnt2 > 0):
        # print(cnt2)
        res += vals[i][1]*vals[i][3]
        cnt2 -= vals[i][1]
        i += 1
       # print(res)
    return res


# filtrar por numero de tortas
l = 0
r = p
while l < r:
    mid = int((l+r)/2)
    ch = h(mid*S)
    nh = h((mid+1)*S)
    if nh - ch > 0:
        l = mid + 1
    else:
        r = mid
print(h(l*S))
